alicloud/resource_alicloud_datahub_subscription.go

```diff
@@ -0,0 +1,191 @@
+			},
+			"create_time": {
+				Type:     schema.TypeString, //uint64 value from sdk
+				Computed: true,
+			},
+			"last_modify_time": {
+				Type:     schema.TypeString, //uint64 value from sdk
+				Computed: true,
+			},
+			"state": {
+				Type:         schema.TypeInt,
+				Optional:     true,
+				ValidateFunc: validateIntegerInRange(0, 3),
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+					return new == strconv.Itoa(types.SUB_CLEANED.Value())
+				},
+			},
+			"is_owner": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAliyunDatahubSubscriptionCreate(d *schema.ResourceData, meta interface{}) error {
+	dh := meta.(*AliyunClient).dhconn
+
+	projectName := d.Get("project_name").(string)
+	topicName := d.Get("topic_name").(string)
+	subComment := d.Get("comment").(string)
+
+	subId, err := dh.CreateSubscription(projectName, topicName, subComment)
+	if err != nil {
+		d.SetId("")
+		return fmt.Errorf("failed to create subscription to '%s/%s' with error: %s", projectName, topicName, err)
+	}
+
+	d.SetId(fmt.Sprintf("%s%s%s%s%s", projectName, COLON_SEPARATED, topicName, COLON_SEPARATED, subId))
+	return resourceAliyunDatahubSubscriptionUpdate(d, meta)
+}
+
+func parseId3(d *schema.ResourceData, meta interface{}) (projectName, topicName, subId string, err error) {
+	split := strings.Split(d.Id(), COLON_SEPARATED)
+	if len(split) != 3 {
+		err = fmt.Errorf("you should use resource alicloud_datahub_subscription's new field 'project_name' and 'topic_name' to re-import this resource.")
+		return
+	} else {
+		projectName = split[0]
+		topicName = split[1]
+		subId = split[2]
+	}
+	return
+}
+
+func resourceAliyunDatahubSubscriptionRead(d *schema.ResourceData, meta interface{}) error {
+	projectName, topicName, subId, err := parseId3(d, meta)
+	if err != nil {
+		return err
+	}
+
+	dh := meta.(*AliyunClient).dhconn
+
+	sub, err := dh.GetSubscription(projectName, topicName, subId)
+	if err != nil {
+		d.SetId("")
+		return fmt.Errorf("failed to get subscription %s with error: %s", subId, err)
+	}
+
+	d.Set("project_name", projectName)
+	d.Set("topic_name", sub.TopicName)
+	d.Set("sub_id", sub.SubId)
+	d.Set("comment", sub.Comment)
+	d.Set("create_time", utils.Uint64ToTimeString(sub.CreateTime))
+	d.Set("last_modify_time", utils.Uint64ToTimeString(sub.LastModifyTime))
+	d.Set("is_owner", sub.IsOwner)
+	d.Set("state", sub.State.Value())
+	return nil
+}
+
+func resourceAliyunDatahubSubscriptionUpdate(d *schema.ResourceData, meta interface{}) error {
+	projectName, topicName, subId, err := parseId3(d, meta)
+	if err != nil {
+		return err
+	}
+
+	dh := meta.(*AliyunClient).dhconn
+
+	if d.HasChange("state") && !d.IsNewResource() {
+		subState := d.Get("state").(types.SubscriptionState)
+
+		err := dh.UpdateSubscriptionState(projectName, topicName, subId, subState)
+		if err != nil {
+			return fmt.Errorf("failed to update subscription %s's state  with error: %s", subId, err)
+		}
```

Comment:
`state` is a output parameter and it can not be modified.

Reply:
After thinking twice, I do agree with you on splitting INPUT and OUTPUT into two different variables.

Reply:
If you want to control flag using state, just make `state` be input is ok and there is no need import another new field `new_state`.

Reply:
By the way, terraform does suggest to control the specified resource status and it expect the resource status is a final stable status, like `Running` or `Available`.